### PR TITLE
CMS: add support for linkable images on program pages

### LINKED ISSUE #4542: CMS: add support for linkable images on program pages
As a Course team member, I'd like to be able to add images to a program page and link them to another page/site.

┆Attachments: <a href="https://trello.com/1/cards/5e664c37e84afb8320536a9c/attachments/5e832e78d384d9078e36bc4d/download?backingUrl=https%3A%2F%2Ftrello-attachments.s3.amazonaws.com%2F5e25a81abbf63943874ebde7%2F5e664c37e84afb8320536a9c%2Ffb0ad323a4567c200bdd93e3eb76367c%2Fimage.png">chat with Wagtail support</a> | <a href="https://share.vidyard.com/watch/cBRHMqcHVzHjWGm4SFFLsT?">screencast</a> | <a href="https://wagtailcms.slack.com/archives/C81FGJR2S/p1583329115133400">link to Wagtail Slack</a> | <a href="https://github.com/mitodl/micromasters/pull/4584">#4584 CMS: add support for linkable images on program pages</a>

### PR DESCRIPTION
**Pull request by asadiqbal08**
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested
#### What are the relevant tickets?
[Ticket](https://trello.com/c/TiX5WMh9/35-cms-add-support-for-linkable-images-on-program-pages)

#### What's this PR do?
Add support to add clickable images in Program and Benefit pages. 

fixes #4542

#### How should this be manually tested?
Follow the shared links in the ticket.

┆Attachments: <a href="https://trello.com/1/cards/5e84960da13dc072e242519a/attachments/5e8dbb926dc39e789ae6efe5/download?backingUrl=https%3A%2F%2Ftrello-attachments.s3.amazonaws.com%2F5e25a81abbf63943874ebde7%2F5e84960da13dc072e242519a%2Ffcd28fee0238180976e8727ca9cc51fc%2Fimage.png">image.png</a>

### CODE REVIEW

**Comment by pdpinch** [comment]
Can we get some screenshots of what the new CMS will look like? And/or a PR build that I can try out? 

Are there any concerns about migrating data on existing pages? 

Small thing: please add "fixes #4542" to the PR description, so this will close the related issue when it merges. This can be instead of the link to Trello.

**Comment by asadiqbal08** [comment]
@pdpinch I attached a video demo under Trello Ticket. did you get a chance to look over it ? I will add the screenshot too.

File: cms/migrations/0048_convert_richtextfield_to_streamfield.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+from wagtail.core.rich_text import RichText
+
+
+def convert_to_streamfield(apps, schema_editor):
+    ProgramPage = apps.get_model("cms", "ProgramPage")
+    for page in ProgramPage.objects.all():
+        if page.description.raw_text and not page.description:
+            page.description = [('paragraph', RichText(page.description.raw_text))]
+            page.save()
```
Comment (asadiqbal08) [inline cms/migrations/0048_convert_richtextfield_to_streamfield.py:16]: @pdpinch regarding your concern for existing pages.

**Comment by pdpinch** [comment]
Is this migration reversible? 

> @pdpinch <https://github.com/pdpinch> regarding your concern for existing pages.

**Comment by asadiqbal08** [comment]
> Is this migration reversible?
> […](#)
> @pdpinch <https://github.com/pdpinch> regarding your concern for existing pages.

Yes. Added both.

**Comment by asadiqbal08** [comment]
@ahmed-arbisoft need review over it.

File: cms/migrations/0048_convert_richtextfield_to_streamfield.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+from django.db import migrations
+import wagtail.core.blocks
+import wagtail.core.fields
+import wagtail.images.blocks
+
+from wagtail.core.rich_text import RichText
+
+
+def convert_to_streamfield(apps, schema_editor):
```
Comment (ahmed-belal) [inline cms/migrations/0048_convert_richtextfield_to_streamfield.py:11]: This is only going to convert the current version of a published page and not update a previous revision, in which case if someone reverts to a previous revision the streamfield will break. Please refer to https://docs.wagtail.io/en/v2.8/topics/streamfield.html#migrating-richtextfields-to-streamfield for instructions on updating all the revisions of published/unpublished pages.

File: cms/templates/cms/imagewithlink.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{% load wagtailimages_tags %}
+
+<a href="{{ self.link }}" target="_blank">
+
```
Comment (ahmed-belal) [inline cms/templates/cms/imagewithlink.html:4]: nit: Could you get rid of the unnecessary empty lines?

File: cms/blocks.py [hunk 4]
```diff
@@ -19,3 +19,18 @@ class CourseTeamBlock(blocks.StructBlock):
...
     )
+
+
+class ImageWithLinkBlock(blocks.StructBlock):
+    """ Image with a clickable link on it """
+    image = ImageChooserBlock(label="Image", required=True, help_text="The image to display.")
+    link = blocks.URLBlock(
+        label="Link",
+        required=True,
+        help_text="Include https:// or http://"
```
Comment (ahmed-belal) [inline cms/blocks.py:30]: nit: `Absolute URL to the image, like https://example.com/some_image.jpg`

File: cms/templates/cms/imagewithlink.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{% load wagtailimages_tags %}
+
+<a href="{{ self.link }}" target="_blank">
+    
+    {% image self.image original %}
```
Comment (ahmed-belal) [inline cms/templates/cms/imagewithlink.html:5]: We're always loading the original image rendition which may not be desirable every time. I do not see any CSS to make this full column width?

Comment (asadiqbal08) [reply cms/templates/cms/imagewithlink.html:5]: It is not already in the current design. not sure either required.

File: cms/migrations/0048_convert_richtextfield_to_streamfield.py (hunk 2 above)
Comment (asadiqbal08) [reply cms/migrations/0048_convert_richtextfield_to_streamfield.py:11]: @pdpinch  a quick question, do we have any draft Program/Benefit pages over PROD ?

**Comment by pdpinch** [comment]
I just checked and there seems to be a lot more drafts than I would have expected. 

What happens with the migration with draft pages?

**Comment by asadiqbal08** [comment]
Looking around the wagtail documentation, get some idea, will update soon.

**Comment by asadiqbal08** [comment]
@pdpinch let me know further improvements as per your requirements. 
Here is the [Demo-replica-from-RC](https://micromasters-ci-pr-4584.herokuapp.com/supply-chain-management/)

**Comment by asadiqbal08** [comment]
@pdpinch Can you take a look over the image alignment option now [here](https://micromasters-ci-pr-4584.herokuapp.com/supply-chain-management/) ? Also take a look over [CMS side](https://micromasters-ci-pr-4584.herokuapp.com/cms/pages/3/) for alignment option while uploading an image. 
Please feel free to share any feedback if need any further improvements in terms of usability.

**Pull request merged**